src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -414,6 +417,29 @@ public Iterator<? extends PrimaryKey> buildIterator(Plan plan)
         }
     }
 
+    /**
+     * Creates an iterator over keys of rows that match given WHERE predicate.
+     * Does not cache the iterator!
+     */
+    private RangeIterator buildIterator(Expression predicate)
+    {
+        QueryView view = getQueryView(predicate.context);
+        return TermIterator.build(predicate, view.referencedIndexes, mergeRange, queryContext, false, Integer.MAX_VALUE);
+    }
+
+    /**
+     * Creates a consistent view of indexes.
+     * Invocations are memorized - multiple calls for the same context return the same view.
+     * The views are kept for the lifetime of this {@code QueryController}.
+     */
+    QueryView getQueryView(IndexContext context)
+    {
+        return queryViews.computeIfAbsent(context,
+                                          c -> new QueryView.Builder(c, mergeRange, queryContext).build());
+
+    }
```

Comment:
When I run the `VectorTypeTest`, I am failing with a leaked reference that passes through this logic. I see you added code that should close the views, but is there any chance it is missed in the vector code path or if the merge from main broke it?

```
ERROR [Reference-Reaper] 2024-10-03 21:37:10,481 Ref.java:276 - Allocate trace org.apache.cassandra.utils.concurrent.Ref$State@582ea780:
Thread[main,5,main]
	at java.base/java.lang.Thread.getStackTrace(Thread.java:1602)
	at org.apache.cassandra.utils.concurrent.Ref$Debug.<init>(Ref.java:266)
	at org.apache.cassandra.utils.concurrent.Ref$State.<init>(Ref.java:188)
	at org.apache.cassandra.utils.concurrent.Ref.<init>(Ref.java:111)
	at org.apache.cassandra.utils.concurrent.Ref.tryRef(Ref.java:153)
	at org.apache.cassandra.io.sstable.format.SSTableReader.tryRef(SSTableReader.java:2282)
	at org.apache.cassandra.utils.concurrent.Refs.tryRef(Refs.java:212)
	at org.apache.cassandra.db.ColumnFamilyStore.selectAndReference(ColumnFamilyStore.java:1983)
	at org.apache.cassandra.index.sai.plan.QueryView$Builder.build(QueryView.java:129)
	at org.apache.cassandra.index.sai.plan.QueryController.lambda$getQueryView$5(QueryController.java:438)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
	at org.apache.cassandra.index.sai.plan.QueryController.getQueryView(QueryController.java:437)
	at org.apache.cassandra.index.sai.plan.QueryController.getTopKRows(QueryController.java:529)
	at org.apache.cassandra.index.sai.plan.QueryController.getTopKRows(QueryController.java:101)
	at org.apache.cassandra.index.sai.plan.Plan$AnnIndexScan.execute(Plan.java:1335)
	at org.apache.cassandra.index.sai.plan.QueryController.buildIterator(QueryController.java:409)
	at org.apache.cassandra.index.sai.plan.StorageAttachedIndexSearcher.search(StorageAttachedIndexSearcher.java:125)
	at org.apache.cassandra.db.ReadCommand.searchStorage(ReadCommand.java:516)
	at org.apache.cassandra.db.ReadCommand.executeLocally(ReadCommand.java:413)
	at org.apache.cassandra.db.ReadCommand.executeInternal(ReadCommand.java:541)
	at org.apache.cassandra.cql3.statements.SelectStatement.executeInternal(SelectStatement.java:613)
	at org.apache.cassandra.cql3.statements.SelectStatement.executeLocally(SelectStatement.java:596)
	at org.apache.cassandra.cql3.statements.SelectStatement.executeLocally(SelectStatement.java:103)
	at org.apache.cassandra.cql3.QueryProcessor.executeInternal(QueryProcessor.java:489)
	at org.apache.cassandra.cql3.CQLTester.executeFormattedQuery(CQLTester.java:1400)
	at org.apache.cassandra.cql3.CQLTester.execute(CQLTester.java:1388)
	at org.apache.cassandra.index.sai.cql.VectorTypeTest.lambda$testEnsureIndexQueryableAfterTransientFailure$16(VectorTypeTest.java:1209)
	at org.assertj.core.api.ThrowableAssert.catchThrowable(ThrowableAssert.java:63)
	at org.assertj.core.api.AssertionsForClassTypes.catchThrowable(AssertionsForClassTypes.java:892)
	at org.assertj.core.api.Assertions.catchThrowable(Assertions.java:1365)
	at org.assertj.core.api.Assertions.assertThatThrownBy(Assertions.java:1209)
	at org.apache.cassandra.index.sai.cql.VectorTypeTest.lambda$testEnsureIndexQueryableAfterTransientFailure$17(VectorTypeTest.java:1209)
	at org.apache.cassandra.cql3.CQLTester.beforeAndAfterFlush(CQLTester.java:1929)
	at org.apache.cassandra.index.sai.cql.VectorTypeTest.testEnsureIndexQueryableAfterTransientFailure(VectorTypeTest.java:1207)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)
```

Reply:
This is happening on main to me as well.